File: phpwrapper/mobileworks.php
<?php


class MobileWorks {

	private $credentials = '';
	const PRODUCTION = false;

	private $task_url, $job_url;

	function __construct( $username, $password ) {
		if ( !extension_loaded( 'curl' ) ) {
			throw new Exception( 'CURL is not installed!' );
		}
		$this->credentials = "$username:$password";
		if ( self::PRODUCTION ) {
			$this->task_url = 'https://work.mobileworks.com/api/v2/task/';
			$this->job_url = 'https://work.mobileworks.com/api/v2/job/';
		}
		else {
			$this->task_url = 'https://staging.mobileworks.com/api/v2/task/';
			$this->job_url = 'https://staging.mobileworks.com/api/v2/job/';
		}
	}

	/**
	 * Creates and sends a request to the specified $url using the specified $method.
	 * @param string $url The URL of the request.
	 * @param string $method The method of the request (GET/POST/DELETE).
	 * @param null|string|array $postData  The data to be posted (optional).
	 * @return array The array representation of the returned JSON object.
	 * @throws Exception
	 */
	function makeRequest( $url, $method = 'GET', $postData = null ) {
		$req = curl_init( $url );
		curl_setopt( $req, CURLOPT_CUSTOMREQUEST, $method );
		if ( $method == 'POST' && !is_null( $postData ) ) {
			curl_setopt( $req, CURLOPT_POSTFIELDS, $postData );
		}
		curl_setopt( $req, CURLOPT_RETURNTRANSFER, true );
		curl_setopt( $req, CURLOPT_HTTPAUTH, CURLAUTH_BASIC );
		curl_setopt( $req, CURLOPT_USERPWD, $this->credentials );

		$result = curl_exec( $req );
		if ( curl_errno( $req ) ) {
			curl_close( $req );
			throw new Exception( curl_error( $req ) );
		}
		$httpCode = curl_getinfo( $req, CURLINFO_HTTP_CODE );
		echo $httpCode, PHP_EOL;
		curl_close( $req );
		if ( $httpCode >= 500 ) {
			throw new Exception( 'A server error occured with the HTTP code: ' . $httpCode );
		}
		if ( $httpCode >= 300 ) {
			throw new Exception( "$httpCode: $result" );
		}
		return get_object_vars( json_decode( $result ) );
	}

	/**
	 * Posts a task to MobileWorks and returns the URL of the created task.
	 * @param array $taskParams The parameters of the task to be created.
	 * @return string The URL of the created task.
	 */
	function postTask( $taskParams ) {
		$response = $this->makeRequest( $this->task_url, 'POST', json_encode( $taskParams ) );
		return $response['Location'];
	}

	/**
	 * Gets the information of the task located in $taskUrl.
	 * @param string $taskUrl The URL of the task to be retrieved.
	 * @return array The task's information.
	 */
	function retrieveTask( $taskUrl ) {
		return $this->makeRequest( $taskUrl );
	}

	/**
	 * Deletes the task located in $taskUrl.
	 * @param string $taskUrl The URL of the task to be deleted.
	 * @return array
	 */
	function deleteTask( $taskUrl ) {
		return $this->makeRequest( $taskUrl, 'DELETE' );
	}

	/**
	 * Posts a job to MobileWorks and returns the URL of the created job.
	 * @param array $jobParams The parameters of the job to be created.
	 * @return string The URL of the created job.
	 */
	function postJob( $jobParams ) {
		$response = $this->makeRequest( $this->job_url, 'POST', json_encode( $jobParams ) );
		return $response['Location'];
	}

	/**
	 * Gets the information of the job located in $jobUrl.
	 * @param string $jobUrl The URL of the job to be retrieved.
	 * @return array The job's information.
	 */
	function retrieveJob( $jobUrl ) {
		return $this->makeRequest( $jobUrl );
	}

	/**
	 * Deletes the job located in $jobUrl.
	 * @param string $jobUrl The URL of the job to be deleted.
	 * @return array
	 */
	function deleteJob( $jobUrl ) {
		return $this->makeRequest( $jobUrl, 'DELETE' );
	}

}

//$mw = new MobileWorks( 'prayag', 'root' );
//$jobParams = array(
//	'instructions' => 'some instructions',
//	'fields' => array(
//		array( 'Name' => 't' ),
//		array( 'Admin Email' => 'e' ),
//		array( 'Alexa Rank' => 'n' ),
//	),
//	'tasks' => array(
//		array( 'resource' => 'http://www.google.com/' ),
//		array( 'resource' => 'http://www.facebook.com/' ),
//		array( 'resource' => 'http://www.youtube.com/' ),
//	),
//);
//$job = $mw->postJob( $jobParams );
//print $job;

//print_r( $mw->retrieveJob( 'https://staging.mobileworks.com/api/v2/job/33/' ) );
//print_r( $mw->deleteJob( 'https://staging.mobileworks.com/api/v2/job/33/' ) );

//print_r( $mw->retrieveTask( 'https://staging.mobileworks.com/api/v2/task/105/' ) );
//print_r( $mw->deleteTask( 'https://staging.mobileworks.com/api/v2/task/89/' ) );

File: python/MobileWorks/mobileworks.py
import urllib2, json, base64


class _API:
    """
    This is the base class for Task/Job.
    It contains the general methods for making API calls to MobileWorks
    """
    
    class Request( urllib2.Request ):
        """
        This class is an extension of urllib2.Request that allows requests other than GET/POST.
        """
        
        def __init__( self, url, data = None, headers = {},
                     origin_req_host = None, unverifiable = False, method = None ):
            """
            These parameters (except `method`) are the same as the parameters in the parent class `urllib2.Request`, which can be found here:
            http://docs.python.org/library/urllib2.html#urllib2.Request
            
            The `method` parameter was added to allow HTTP requests other than GET/POST.
            """
            self._method = method
            urllib2.Request.__init__( self, url, data, headers, origin_req_host, unverifiable )
        
        def get_method( self ):
            return self._method if self._method else urllib2.Request.get_method( self )
        
    @classmethod
    def _make_request( cls, url, method = None, post_data = None ):
        """
        Creates and sends an HTTP request.
        """
        if MobileWorks.credentials is None:
            raise Exception( 'You are not logged in.' )
        req = cls.Request( url, method = method, data = post_data, headers = {'Content-Type':'application/json'} )
        req.add_header( 'Authorization', 'Basic ' + MobileWorks.credentials )
        
        try:
            response = urllib2.urlopen( req )
            headers = response.info().dict
            content = response.read()
            response.close()
            return headers, content
#            return json.loads( content )
        except urllib2.HTTPError, e:
            if e.code >= 500:
                raise Exception( 'HTTP %d: A server error occured' % e.code )
            else:
                raise Exception( 'HTTP %d: %s' % ( e.code, e.read() ) )
    
    VERSION = 2
    location = None
    
    def url( self ):
        return MobileWorks.DOMAIN + self._path()
    
    def dict( self ):
        """
        This is used for json serialization.
        It should be overriden by subclasses.
        """
        return self.__dict__
    
    def to_json(self):
        return json.dumps( self.dict() )
    
    def post( self ):
        """
        Posts the object to MobileWorks API and returns the URL of the created object.
        """
        headers, content = self._make_request( self.url(), 'POST', self.to_json() )
        if self.VERSION == 1:
            self.location = headers['location']
        elif self.VERSION == 2:
            self.location = json.loads(content)['Location']
        return self.location
    
    @classmethod
    def _retrieve( cls, url ):
        """
        Gets the information of the object located at `url`.
        """
        headers, content = cls._make_request( url )
        return json.loads( content )
    
    def delete( self ):
        """
        Deletes the object located at `url`.
        """
        if self.location is None:
            raise Exception( "This object doen't point to any resource on the server." )
        headers, content = self._make_request( url, 'DELETE' )
        if cls.VERSION == 1:
            return True
        elif cls.VERSION == 2:
            return json.loads( content )
    

class Task(_API):
    
    def __init__( self, location = None, **task_params ):
        if location is not None and task_params:
            raise Exception( "You can't provide a location and task parameters at the same time" )
        if location is not None:
            self.location = location
            self._load()
        else:
            self.params = task_params
            self.fields = None
        
    def _load( self ):
        self.params = Task._retrieve( self.location )
        if 'fields' in self.params:
            self.fields = self.params['fields']
            del self.params['fields']
        else:
            self.fields = None
            
    def _path( self ):
        """
        Returns the base path of tasks depending on the API version.
        """
        if self.VERSION == 1:
            return 'api/v1/tasks/'
        elif self.VERSION == 2:
            return 'api/v2/task/'
        raise Exception( 'Sorry, version %d is not supported by the library yet!' % self.VERSION )
        
    def get_param( self, name ):
        """
        Gets the specified parameter from this task.
        """
        return self.params[name]
        
    def set_params( self, **params ):
        """
        Sets parameters of this task.
        """
        self.params.update( params )
    
    def add_field( self, name, type, **kwargs ):
        """
        Adds a field to this task.
        """
        if self.VERSION < 2:
            raise Exception( "Fields only exist in version 2 of the API" )
        
        if self.fields is None:
            self.fields = []
            
        field = {name: type}
        if kwargs:
            field.update( kwargs )
        self.fields.append( field )
        
    def dict( self ):
        dic = self.params.copy()
        if self.fields is not None:
            dic.update( {'fields': self.fields} )
        return dic
        
    
class Job(_API):
    
    def __init__( self, location = None, **job_params ):
        if location is not None and job_params:
            raise Exception( "You can't provide a location and job parameters at the same time" )
        if location is not None:
            self.location = location
            self._load()
        else:
            self.params = job_params
            self.tasks = []
            self.fields = None
        
    def _load( self ):
        self.params = Job._retrieve( self.location )
        if 'tasks' in self.params:
            self.tasks = self.params['tasks']
            del self.params['tasks']
        else:
            self.tasks = []
            
        if 'fields' in self.params:
            self.fields = self.params['fields']
            del self.params['fields']
        else:
            self.fields = None
        
    def _path( self ):
        """
        Returns the base path of jobs depending on the API version.
        """
        if self.VERSION == 1:
            return 'api/v1/job/'
        elif self.VERSION == 2:
            return 'api/v2/job/'
        raise Exception( 'Sorry, version %d is not supported by the library yet!' % self.VERSION )
        
    def get_param( self, name ):
        """
        Gets the specified parameter from this job.
        """
        return self.params[name]

    def set_params( self, **params ):
        """
        Sets parameters of this job.
        """
        self.params.update( params )
        
    def add_task( self, task ):
        """
        Adds a task to this job.
        """
        if task.__class__ == Task:
            self.tasks.append( task )
        else:
            raise ValueError( "`task` must be an instance of the Task class" )
    
    def add_field( self, name, type, **kwargs ):
        """
        Adds a field to this job.
        """
        if self.VERSION < 2:
            raise Exception( "Fields only exist in version 2 of the API" )
        
        if self.fields is None:
            self.fields = []
            
        field = {name: type}
        if kwargs:
            field.update( kwargs )
        self.fields.append( field )
    
    def dict( self ):
        dic = self.params.copy()
        if self.fields is not None:
            dic.update( {'fields': self.fields} )
        if self.tasks is not None:
            tasks = self.tasks
            if len( self.tasks ) and self.tasks[0].__class__ == Task:
                tasks = [t.dict() for t in tasks]
            dic.update( {'tasks': tasks} )
        return dic
    

class MobileWorks:
    """
    This class is used to login to MobileWorks and keeps track of the user credentials and
    the DOMAIN used for the API. 
    """
    
    credentials = None
    DOMAIN = 'https://work.mobileworks.com/'
    PROFILE_PATH = 'api/v1/userprofile/'
    
    @classmethod
    def login( cls, username, password ):
        cls.credentials = base64.encodestring( username + ':' + password )[:-1]
        try:
            headers, content = _API._make_request(cls.DOMAIN + cls.PROFILE_PATH)
            return json.loads( content )['objects'][0]
#            return _API._make_request(cls.DOMAIN + cls.PROFILE_PATH)['objects'][0]
        except Exception, e:
            print e
            cls.credentials = None
            raise Exception( "Login failed! To reset your password, please go to '%saccounts/password_reset/'" % cls.DOMAIN )
    
        
    @classmethod
    def local(cls, port = 8000):
        cls.DOMAIN = 'http://localhost:%d/' % port
    
    @classmethod    
    def staging(cls):
        cls.DOMAIN = 'https://staging.mobileworks.com/'
        
    @classmethod
    def sandbox(cls):
        cls.DOMAIN = 'https://sandbox.mobileworks.com/'
        
    @classmethod
    def production(cls):
        cls.DOMAIN = 'https://work.mobileworks.com/'
        
    @classmethod
    def version( cls, v = None ):
        if v == None:
            return _API.VERSION
        elif type( v ) == int:
            _API.VERSION = v
        else:
            raise Exception( 'Version must be an integer!' )
